### PR TITLE
Introduce transaction pool interface

### PR DESCRIPTION
**Pull request by fab-10**
Signed-off-by: Fabio Di Fabio <fabio.difabio@consensys.net>

## PR description

Preparatory work to introduce more implementations of the transaction pool, specifically the new transaction pool that will support 4844.

No functionality changes in this PR, apart the introduction of the interface, sorter unit tests have been moved to the package of the corresponding classes

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by Gabriel-Trintinalia — APPROVED** [review]
LGTM!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java [hunk 1]
```diff
@@ -65,10 +66,8 @@
 /**
  * Holds the current set of pending transactions with the ability to iterate them based on priority
  * for mining or look-up by hash.
- *
- * <p>This class is safe for use across multiple threads.
```
Comment (jframe) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java:69]: Did something change to make this no longer thread safe?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/GasPricePendingTransactionsTest.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 public class GasPricePendingTransactionsTest extends AbstractPendingTransactionsTestBase {
 
   @Override
-  AbstractPendingTransactionsSorter getSorter(
+  PendingTransactions getSorter(
```
Comment (jframe) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/GasPricePendingTransactionsTest.java:28]: rename to getPendingTransaction?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolFactory.java [hunk 3]
```diff
@@ -172,7 +171,7 @@ private static void subscribeTransactionHandlers(
         .subscribe(EthPV65.NEW_POOLED_TRANSACTION_HASHES, pooledTransactionsMessageHandler);
   }
 
-  private static AbstractPendingTransactionsSorter createPendingTransactionsSorter(
+  private static PendingTransactions createPendingTransactionsSorter(
```
Comment (jframe) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolFactory.java:174]: name isn't quite right anymore. Maybe createPendingTransactions?

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java [hunk 4]
```diff
@@ -56,7 +56,7 @@
 @RunWith(MockitoJUnitRunner.class)
 public class MergeReorgTest implements MergeGenesisConfigHelper {
 
-  @Mock AbstractPendingTransactionsSorter mockSorter;
+  @Mock PendingTransactions mockSorter;
```
Comment (jframe) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java:59]: rename to mockPendingTransactions

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/PendingTransactions.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/PendingTransactions.java:2]: nit: Contributors to Hyperledger Besu

File: ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockTransactionSelectorTest.java [hunk 6]
```diff
@@ -94,7 +94,7 @@ public void setup() {
     when(transactionValidator.getGoQuorumCompatibilityMode()).thenReturn(true);
   }
 
-  protected abstract AbstractPendingTransactionsSorter createPendingTransactionsSorter();
+  protected abstract PendingTransactions createPendingTransactionsSorter();
```
Comment (jframe) [inline ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockTransactionSelectorTest.java:97]: rename to createPendingTransactions

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/PendingTransactions.java (hunk 5 above)
Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/PendingTransactions.java:2]: fixed

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java (hunk 1 above)
Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java:69]: No, removed by error and readded

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/GasPricePendingTransactionsTest.java (hunk 2 above)
Comment (fab-10) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/GasPricePendingTransactionsTest.java:28]: thanks for suggesting the renaming, it make sense and I have applied them all

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolFactory.java (hunk 3 above)
Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolFactory.java:174]: done

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java (hunk 4 above)
Comment (fab-10) [reply consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java:59]: done

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java [hunk 7]
```diff
@@ -88,7 +88,7 @@ public void setUp() {
             protocolContext,
             mockProtocolSchedule,
             CompletableFuture::runAsync,
-            mockSorter,
+            mockPeningTransactions,
```
Comment (macfarla) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java:91]:
```suggestion
            mockPendingTransactions,
```

**Review by macfarla — APPROVED** [review]
LGTM. one tiny typo

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java [hunk 8]
```diff
@@ -56,7 +56,7 @@
 @RunWith(MockitoJUnitRunner.class)
 public class MergeReorgTest implements MergeGenesisConfigHelper {
 
-  @Mock AbstractPendingTransactionsSorter mockSorter;
+  @Mock PendingTransactions mockPeningTransactions;
```
Comment (macfarla) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeReorgTest.java:59]:
```suggestion
  @Mock PendingTransactions mockPendingTransactions;
```

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java (hunk 1 above)
Comment (macfarla) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java:69]: checked it's readded.

**Pull request merged**